File: DoitMission_12/Doit12_Java/app/src/main/java/com/example/doit12_java/MainActivity.java
package com.example.doit12_java;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {

    public static final String name = "NAME";
    public static final String message = "MESSGAE";

    private TextReceiver textReceiver = new TextReceiver();

    private Button btn;
    private EditText edit;
    private TextView text;

    @Override
    protected void onResume() {
        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addAction("BroadCast");
        registerReceiver(textReceiver,intentFilter);
        super.onResume();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        btn = findViewById(R.id.sendbutton);
        edit = findViewById(R.id.editText);
        text = findViewById(R.id.receiveText);

        btn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String editText = edit.getText().toString();
                Intent serviceIntent = new Intent(getApplicationContext(),TextService.class);
                serviceIntent.putExtra(name,"Service");
                serviceIntent.putExtra(message,editText);
                startService(serviceIntent);
            }
        });

        Intent intent = getIntent();
        process(intent);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        process(intent);
        super.onNewIntent(intent);
    }

    private void process(Intent intent){
        if(intent.getStringExtra(message) != null){
            text.setText(intent.getStringExtra(message));
        }
    }

    @Override
    protected void onDestroy() {
        unregisterReceiver(textReceiver);
        super.onDestroy();
    }
}

File: DoitMission_12/Doit12_Kotlin/app/src/main/java/com/example/doit12_kotlin/MainActivity.kt
package com.example.doit12_kotlin

import android.content.BroadcastReceiver
import android.content.Intent
import android.content.IntentFilter
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    companion object {
        const val name = "NAME"
        const val message = "MESSAGE"
    }

    lateinit var receiver: BroadcastReceiver

    override fun onResume() {
        receiver = TextReceiver()
        val intentFilter = IntentFilter()
        intentFilter.addAction("BroadCast")
        registerReceiver(receiver,intentFilter)
        super.onResume()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        sendbutton.setOnClickListener {
            val serviceIntent = Intent(applicationContext, TextService::class.java)
            serviceIntent.putExtra(name,"Service")
            serviceIntent.putExtra(message,editText.text.toString())
            Log.e("Start","서비스 시작 전 호출")
            startService(serviceIntent)
        }

        val mainIntent = intent
        process(mainIntent)
    }

    private fun process(intent : Intent){
        if(intent.getStringExtra(message) != null) {
            receiveText.text = intent.getStringExtra(message)
        }
    }

    override fun onNewIntent(intent: Intent) {
        process(intent)
        super.onNewIntent(intent)
    }

    override fun onDestroy() {
        unregisterReceiver(receiver)
        super.onDestroy()
    }

}
